[WIP] Particle property pool

gassmoeller opened the pull request:
A first try for #1215. It kind of works (no more segfaults and accesses to already freed memory) but it is currently not scalable (the translation table just grows over time until it is too large to handle). I have two questions regarding the implementation:
- Should a particle upon destruction deregister its memory? This would mean for every temporary copy of particles we would need to register/free memory in the pool. The alternative would be to allow several particles to keep a handle to the same properties (of course only if one of them is a temporary copy)
- To consolidate the translation_table I would need to change the handles of all particles. To do that I would need to pass the particles map to the property pool. But then I run into a kind of circular include problem: ParticlePool needs to know what a `types::LevelInd` and a `Particle` is, while `Particle` needs to know what a `ParticlePool` and a `ParticlePool::Handle` is. Normally I would resolve this by forward declarations, but I can not forward declare the handle, because it is a member struct of a forward declared class, and I can not forward declare the typedef LevelInd. I could move `Handle` out of the `ParticlePool` class, would that be an acceptable solution, or do I miss some easy way to resolve this?

bangerth commented:
1/ Yes, particles that are destroyed need to release memory. That means that we need to keep a free list for the pool so we can re-use this memory. (If you do temporary copies, you can avoid the allocation/deallocation if you implemented a C++11-style move constructor or move assignment operator. That will of course only be efficient with C++11, but I have to admit that I don't particularly care otherwise.)

2/ No that's the wrong approach. The handles are really just indices, and what they reference must stay entirely within the pool object. Let me look at the code and I may be able to comment more.

bangerth commented on include/aspect/particle/property/interface.h:
```diff
@@ -623,6 +633,11 @@ namespace aspect
+           * This object owns and organizes the memory for all particle
+           * properties.
+           */
+          std_cxx11::unique_ptr<PropertyPool> property_pool;
```
Do you make this a pointer instead of a regular member variable because you don't know the number of property components yet at the time you construct the enclosing object? If so, maybe add this information to the comment.

bangerth commented on include/aspect/particle/property_pool.h:
```diff
@@ -0,0 +1,100 @@
+            index(index)
+          {}
+
+          Handle &operator= (const Handle &handle)
```
Isn't this the default implementation? Do you need to actually implement it?

bangerth commented on include/aspect/particle/property_pool.h:
```diff
@@ -0,0 +1,100 @@
+            index(numbers::invalid_unsigned_int)
+          {}
+
+          Handle(const std::size_t index)
```
I think it would be good to use a typedef in class `PropertyPool` instead of a hard-coded `std::size_t`. We may want to use a 32-bit integer, for example, assuming that no processor stores more than 4B particles.

bangerth commented on include/aspect/particle/property_pool.h:
```diff
@@ -0,0 +1,100 @@
+            return index != numbers::invalid_unsigned_int;
+          }
+
+          std::size_t index;
```
can you make it `private`?

bangerth commented on include/aspect/particle/property_pool.h:
```diff
@@ -0,0 +1,100 @@
+        std::vector<double> memory_pool;    // size == N_slots * n_properties_per_slot;
+
+        // possibly a translation table from index to memory pool location
+        std::vector<std::size_t> translation_table; // size == N_slots
```
Two comments here:
1/ I think the correct inner type is probably `std::vector<double>::size_type` because that's what `memory_pool` uses to index into its array. (Likely `std::size_t`, but let's make it explicit.)

1/ To make this efficient, you will need to keep a list of which slots in `memory_pool` are currently used and which are not. You can in principle gather this information by walking `translation_table` and storing which slots you have seen (assuming that all unused indices in `translation_table` are properly invalidated). But this is not efficient if you want to find an unused slot. My thought was to maybe use an `IndexSet` that represents the indices of all slots currently unused. Allocating a slot would then require finding any element of the index set and then deleting this element from the index set (I'm currently not sure whether we support this last step efficiently), whereas de-allocating a slot would require adding this index back into the index set. (I've also thought about keeping the inverse, i.e., the set of slots that are *used* in such an `IndexSet`, but I think that has no advantage.)

bangerth commented on source/particle/property_pool.cc:
```diff
@@ -0,0 +1,133 @@
+        memory_pool.resize(memory_pool.size() + n_properties);
+
+      if (translation_table.size() < (next_free_index+1))
+        translation_table.resize(translation_table.size() + 1);
```
That's really inefficient, because you will resize this all the time. Better multiply the previous size by a fixed factor, say 1.5 or 2.

bangerth commented:
There's a conflict in this patch now. Can you resolve this? Also, any progress? (Not urgent, only asking.)

gassmoeller commented:
I postpone this for a while in order to finish the other stuff, but thanks for the comments! Will look into them sometime.

gassmoeller commented:
I have updated the PR to a working version using deal.II IndexSet's class (only running with https://github.com/gassmoeller/dealii/tree/add_pop_front_index_set). I have added a screenshot of the profiling results of an older master (the base of this branch) on the left compared to this PR on the right. The middle panel of the left column lists all functions in the `Particle` class, and their runtime. Constructors are listed at the top, the destructor at the bottom. For the current master the destructor is not listed (maybe because it is default created?), but it should be on the order of 50-100 cycles per call (one deallocation of std::vector), while the new one takes about 3000 cycles. While the current version reduces the time spent in the Particle constructor by about 30%, it increases the time spend in the destructor in a way that more than offsets all the gains. This is because every call to `deallocate_particle_properties` creates a new entry in the `ranges` vector of the `IndexSet` class, which involves a call to `std::vector<Range>::insert` and a copy of all elements behind this entry. A test using `pop_front` to allocate new particles instead of `pop_back` is even worse (in this case constructor and destructor need to copy large parts of the vector). 

![comparison_pool](https://cloud.githubusercontent.com/assets/7582930/20231755/4d7568c8-a821-11e6-8489-cec6f467a132.png)

bangerth commented:
I'm going to look into the patch here later, but can you submit the three commits on your branch for `IndexSet` separately? These seem like useful functions in general. If you added testcases, they'd be quite useful independently.

bangerth commented:
I need to look at this in more detail, but I think I have an idea for how to deal with this.

bangerth commented on include/aspect/particle/particle.h:
```diff
@@ -166,16 +172,13 @@ namespace aspect
+         * therefore frees that memory space for other particles. (Note:
+         * the memory is managed by the property_pool, and the memory is not
+         * deallocated by this function, it is kept in reserve for other
+         * particles).
```
I would remove this parenthetical comment. It comments on an implementation detail of the property pool, but the property pool may implement an entirely different strategy -- which, however, is not something the particle class will actually care about.

bangerth commented:
Would it be easy to separate out the introduction of `ArrayView`? This seems to be a relatively independent part of the patch that I think could be universally useful, regardless of whether we do or do not want to keep the pool. It also makes us more flexible later on in changing underlying data structures if we want to.

But if this separation is too complicated, then no worries.

bangerth commented on source/particle/property_pool.cc:
```diff
@@ -0,0 +1,100 @@
+    void
+    PropertyPool::deallocate_properties_array (Handle handle)
+    {
+      free_slots.add_index(handle);
```
This line is the problematic and expensive one, right? If you can confirm, I think I can describe a solution.

gassmoeller replied:
Yes, this is the expensive part.

gassmoeller replied:
Ah, I think I know what you are thinking of (or not, but something else that might help). Since we organized the properties in a way that we can move them around, we can swap the properties of the particle that is going to be deleted with the ones from the particle next to a free range (e.g. the large free range at the end of the IndexSet) and then release the properties without creating a new range in the IndexSet. Is this approximately what you were thinking of?

bangerth replied:
Might have to think a bit more what happens if the last slot in the array is already free, but you get the idea.

bangerth commented on source/particle/property_pool.cc:
```diff
@@ -0,0 +1,100 @@
+      Assert(!free_slots.is_element(handle),
+             ExcMessage("The particle property pool was asked to access a "
+                 "region of memory that is allocated for particle properties, "
+                 " but was reported as not in use."));
```
no need for space at the front.

bangerth commented on source/particle/property_pool.cc:
```diff
@@ -0,0 +1,100 @@
+                 "region of memory that is allocated for particle properties, "
+                 " but was reported as not in use."));
+
+      // Here we make use of the fact that Handle is nothing else than an index
```
That's a Germanism :-) Say "...is just an index..." or "...is nothing but..."

bangerth commented on source/particle/property_pool.cc:
```diff
@@ -0,0 +1,100 @@
+          free_slots.clear();
+          free_slots.set_size(size);
+          free_slots.add_range(0,size);
+          free_slots.subtract_set(old_used_slots);
```
I think you could do this in a simpler way:
```
  IndexSet new_free_slots (n_slots);
  new_free_slots.add_range (n_old_slots, n_new_slots);  // all new slots are free
  for (IndexSet::RangeIterator r=...)
    new_free_slots.add_range (*r);  // or similar; a previously free range is still free
```

bangerth commented:
@gassmoeller -- I think you had a newer version of this, but it's not on github. What's the current status?

gassmoeller commented:
I will work on this tomorrow.

bangerth commented:
OK. As discussed today, I think it would be nice to break out the `ArrayView` stuff and get it committed separately. There should be no discussion about the usefulness of this part by itself.

bangerth commented:
I look forward to the rebased version :-)

(I tried to do it myself this morning, but you had a few changes in the other PR that led to merge conflicts and I had to turn to our paper first...)
